File: libs/dmx/dmx.ts
/**
 * DMX protocol functions.
 */
//% weight=100 color="#0078d7" icon="\uf185"
namespace dmx {
    // http://www.dmx512-online.com/dmx512_packet.html
    // http://cdb.s3.amazonaws.com/ItemRelatedFiles/10191/dmx-101-handbook.pdf
    // https://opendmx.net/index.php/DMX512-A
    // DMXMaster schematics http://www.ca.diigiit.com/download/dmx-shield-rev4.pdf

    export class DMXSerial {
        tx: DigitalInOutPin;
        rx: DigitalInOutPin;
        ser: serial.Serial;

        constructor(tx: DigitalInOutPin) {
            this.tx = tx;
            this.ser = serial.createSerial(this.tx, undefined, 12000);
            this.ser.serialDevice.setBaudRate(250000);
        }

        write(channels: Buffer) {
            // https://learn.sparkfun.com/tutorials/introduction-to-dmx
            // http://www.dmx512-online.com/dmx512_packet.html
            // break (low) > >88 microsecs or 22 pulses
            this.tx.digitalWrite(false);
            control.waitMicros(88);
            // mark after break (high) - 8 microsecs or 2 pulses
            this.tx.digitalWrite(true);
            control.waitMicros(8); 
            // start code is channel 0
            // frames - max 512, no mtbf
            this.ser.serialDevice.writeBuffer(channels);
            // no mtbp
        }
    }

    let _sercom: DMXSerial;
    /**
     * Writes a DMX channel buffer
     * @param buf 
     */
    //%
    export function sendBuffer(buf: Buffer) {
        if (!_sercom) {
            const tx = pins.pinByCfg(DAL.CFG_PIN_DMX_OUT);
            if (!tx) return;
            _sercom = new DMXSerial(tx);
        }

        if (_sercom)
            _sercom.write(buf);
    }
}

File: libs/dmx/dmx.cpp
#include "pxt.h"
#include "neopixel.h"


#define DMX_MARK 1
#define DMX_BREAK 0

/**
 * Functions to operate colored LEDs.
 */
//% weight=100 color="#0078d7" icon="\uf00a"
namespace dmx {
    // http://www.dmx512-online.com/dmx512_packet.html
    // http://cdb.s3.amazonaws.com/ItemRelatedFiles/10191/dmx-101-handbook.pdf
    // https://opendmx.net/index.php/DMX512-A
    // 1 bit = 4us
    void edge(DigitalInOutPin pin, uint value, uint us) {
        pin->setDigitalValue(value);
        target_wait_us(us);
    }

    //% parts="dmx"
    void sendBuffer(DigitalInOutPin pin, Buffer buf) {
        const uint8_t* data = buf-> data;
        uint16_t length = min(512, buf->length);

        // break
        edge(pin, DMX_BREAK, 120);
        // mark after break
        edge(pin, DMX_MARK, 12);
        // start code
        edge(pin, DMX_BREAK, 9);
        edge(pin, DMX_MARK, 2);
        // channels
        for (uint16_t i = 1; i < length; ++i) {
            uint8_t channel = data[i];
            // mark between frame
            edge(pin, DMX_MARK, 2);
            // channel data
            // 1 bit start
            edge(pin, DMX_BREAK, 1);
            // 8 bit LE
            for (i = 0xF0; i; i >>= 1)
                edge(pin, (channel & i) != 0, 1);
            // 2 stop bits
            edge(pin, DMX_MARK, 2);
        }
        // mark between packet needed from 0 to 1sec
        edge(pin, DMX_MARK, 0);
        sleep_ms(1);        
    }
}
